Fix recovery codes issue #26104

ChrisLiangwx opened the pull request:
### Fix recovery codes issue #26104
This PR addresses issue #26104 where the recovery codes were not properly validated on the server side after being generated.
#### Changes
Added validation logic in RecoveryAuthnCodesAction to ensure that the recovery codes submitted by the user match the codes generated by the server. 
The server-generated codes are recorded in FreeMarkerLoginFormsProvider when generating the codes.

Closes #26104

ChrisLiangwx committed:
Fix recovery codes issue

Added validation logic in RecoveryAuthnCodesAction to ensure that the recovery codes submitted by the user match the codes generated by the server. The server generated codes are recorded in FreeMarkerLoginFormsProvider when generating the codes.
Add one constant in RecoveryAuthnCodesAction class, which is used by RecoveryAuthnCodesAction and FreeMarkerLoginFormsProvider

Closes #26104

Signed-off-by: Weixiao Liang <ChrisLiangwx@gmail.com>

mposolda commented on services/src/main/java/org/keycloak/authentication/requiredactions/RecoveryAuthnCodesAction.java:
```diff
@@ -98,6 +98,37 @@ public void processAction(RequiredActionContext reqActionContext) {
         generatedUserLabel = httpReqParamsMap.getFirst(FIELD_USER_LABEL_HIDDEN);
 
+        // Retrieve the initially generated recovery codes from the authentication session
+        String storedCodes = reqActionContext.getAuthenticationSession().getAuthNote("initialGeneratedCodes");
```
Minor: Is it possible to introduce some constant (maybe in class `RecoveryAuthnCodesAction` which already contains few constants) and make sure that both places use this constant?

mposolda reviewed (commented):
@ChrisLiangwx Thanks for the PR!

Added one minor note inline. Besides that, it would be good to add automated test for this and check that when using additional/different codes, the error is actually thrown as expected. We have `RecoveryAuthnCodesAuthenticatorTest`, so maybe new test method could be added here?

ChrisLiangwx commented:
> @ChrisLiangwx Thanks for the PR!
> 
> Added one minor note inline. Besides that, it would be good to add automated test for this and check that when using additional/different codes, the error is actually thrown as expected. We have `RecoveryAuthnCodesAuthenticatorTest`, so maybe new test method could be added here?

Thank you for the feedback!

I have added the constant as requested and updated the relevant code to use it.

Regarding the test method, I encountered an issue. The original generated recovery codes are recorded in the FreeMarkerLoginFormsProvider class. When attempting to run the test, I consistently receive the error "org.keycloak.theme.FreeMarkerException: Failed to process template index.ftl." I'm having difficulty locating the correct theme configuration.

If you could provide guidance or resources on configuring the themes correctly for the test environment, it would be greatly appreciated. This would help in ensuring the automated test checks for errors when additional or different codes are used, as expected.

mposolda commented:
@ChrisLiangwx Are you able to run `RecoveryAuthnCodesAuthenticatorTest` when running this test class from your IDE? For instance Intellij idea? If not, the step1 might be to fix this to be able to run the existing tests before development of the new one.

jonkoops commented on services/src/main/java/org/keycloak/authentication/requiredactions/RecoveryAuthnCodesAction.java:
```diff
@@ -81,7 +82,7 @@ public void requiredActionChallenge(RequiredActionContext context) {
 
     @Override
-    public void processAction(RequiredActionContext reqActionContext) {
+        public void processAction(RequiredActionContext reqActionContext) {
```
```suggestion
    public void processAction(RequiredActionContext reqActionContext) {
```

jonkoops commented on services/src/main/java/org/keycloak/authentication/requiredactions/RecoveryAuthnCodesAction.java:
```diff
@@ -98,6 +99,37 @@ public void processAction(RequiredActionContext reqActionContext) {
+        }
+        List<String> initialGeneratedCodes = Arrays.asList(storedCodes.split(","));
+
+        // Check the number of recovery codes
+        if (generatedCodes.size() != initialGeneratedCodes.size()) {
+            Response challenge = reqActionContext.form()
+                    .setError("Invalid number of recovery codes.")
+                    .createErrorPage(Response.Status.BAD_REQUEST);
+            reqActionContext.challenge(challenge);
+            return;
+        }
+
+        // Check the content of the recovery codes
+        for (String code : generatedCodes) {
+            if (!initialGeneratedCodes.contains(code)) {
+                Response challenge = reqActionContext.form()
+                        .setError("Invalid recovery code.")
+                        .createErrorPage(Response.Status.BAD_REQUEST);
+                reqActionContext.challenge(challenge);
+                return;
+            }
+        }
```
Rather than checking these manually, perhaps we can use the `equals()` method of the list to check these. They should always be in the same order AFAIK.


```java
if (!generatedCodes.equals(initialGeneratedCodes)) {
    Response challenge = reqActionContext.form()
            .setError("Submitted recovery codes do not match generated ones.")
            .createErrorPage(Response.Status.BAD_REQUEST);
    reqActionContext.challenge(challenge);
    return;
}
````

jonkoops commented:
I have to be honest, I am kind of confused why we're putting the recovery codes in an HTML form in the first place. Can we not extract this information from the session instead?
